### PR TITLE
Issue 21/multi class multi output close #21

### LINKED ISSUE #21: Support multi-class and multi-output classification
### Description
Currently, the code trains video models for a single outcome (`Outcome`) per sample. However, with our recent dataset from Labelbox, the `Outcome` now contains tuples representing multiple attributes (e.g., `(5,1,1)` indicating a right coronary with stenting and contrast).

**Tasks to Complete:**
1. **Modify Data Processing:**
   - Adapt data loading to handle `Outcome` columns containing tuples.
   - **Example:**
     ```csv
     FileName,Outcome,Target_loc_label
     video1.avi,"(5,1,1)",5
     video2.avi,"(9,0,1)",9
     ```

2. **Adjust the Loss Function:**
   - Implement a loss function suitable for multi-output, such as `MultiLabelSoftMarginLoss` or a combination of individual losses for each attribute.

3. **Update the Neural Network’s Classification Head:**
   - Modify the output layer to predict multiple labels simultaneously.
   - **Example:** For three attributes, the output layer should have three neurons.

4. **Update Training and Evaluation Processes:**
   - Ensure metrics account for multiple outputs.
   - **Example:** Calculate accuracy for each attribute individually and overall.

**Examples:**

- **Current Data Format:**
  ```csv
  FileName,Outcome,Target_loc_label
  video1.avi,5,Right_coronary
  video2.avi,9,Left_coronayr
  
  -**New Data Format**
  FileName,Outcome,Target_loc_label
video1.avi,"(5,1,1)",(right_coronary, stenting, contrast)
video2.avi,"(9,0,1)",(left_coronary, stenting, contrast)
  
-**Suggested changes**
-Change video.py to support Outocme as a tuple  as well as Target_loc_label as a tuple
- Support a new loss `loss: multi_label_loss` in config.yaml and train.py
- Make it so that `num_classes ` can be a tuple (3,1,1) for 3 classes for outcome in position 0, 1 class in position 1 and 1 class in position 2
- Change neural network head
-

### PR DESCRIPTION
**Pull request by JacquesDelfrate**
See context and original suggested solution for #21 

This PR is divided into the following sections:

# Input files structure
This PR proposes an alternative solution regarding the structure of the input data files:
1. One Column per Head/Class:
Instead of embedding multiple classes as a string in a single DataFrame column and decoding them as a tuple, use one column per head/class.
2. Configuration of Head Structure and Labels Mapping: 
Define the head structure and labels mapping within the configuration files (see `config_x3d_binary_classification_rv.yaml` and `config_x3d_multi_output_classification.yaml`).
3. Optional Labels Mapping in Data Files:
Optionally, users can store the labels mapping within their CSV or Parquet files.

This solution enhances the readability of data files, simplifies the decoding system for classes, and ensures the reproducibility of experiments.

# MutiHead
The `MultiHead` class functions as a hashmap that follows the `head_structure` defined by the user in the YAML configuration file. Each element of the dictionary wraps the previously defined `RegressionHead` object with the appropriate number of classes, effectively replacing the previous `classification` task approach.

```yaml
# Head structure configuration
head_structure:
  contrast_agent: !!int 1    # Binary classification (0 or 1)
  main_structure: !!int 5    # Multi-class classification (5 classes)
  stent_presence: !!int 1    # Binary classification (0 or 1)
```

# MultiLoss
Similarly, `MultiLoss` uses a hashmap that follows the `loss_structure` defined in the YAML file. Users can choose which loss function to apply for each head. This PR also refactors the loss components of the code using a decorator-based registry system, adding a layer of abstraction for better modularity and scalability. Currently, only weighted (default 1.0) class multi-loss has been implemented, but `MultiLabelSoftMarginLoss` can be easily added within this structure.

```yaml
# Loss structure configuration
loss_structure:
  contrast_agent: !!str bce_logit
  main_structure: !!str ce
  stent_presence: !!str bce_logit
```

# Wandb logging
The `MultiHead` and `MultiLoss` implementations introduce more variables for plotting on Wandb. Therefore, users can optionally add a `labels_map` structure to the configuration file. This map will be used to forward losses and metrics with the correct names for each head and class, enhancing the understandability of results on Wandb dashboards.

```yaml
# Label mappings
labels_map:
  contrast_agent:
    "yes": !!int 1
    "no": !!int 0
  main_structure:
    Left Coronary: !!int 0
    Right Coronary: !!int 1
    Graft: !!int 2
    Catheter: !!int 3
    Femoral: !!int 4
  stent_presence:
    present: !!int 1
    absent: !!int 0`
```

#  Installation Instructions
Finally package installation using uv is proposed.
```bash
python -m venv orion-venv
source orion-venv/bin/activate
pip install --upgrade pip
pip install uv
uv pip install -r requirements.txt
```

# Legacy Configurations
All previously config files has been kept for now and moved to notebooks/config/legacy folder.

## Summary by Sourcery

Implement multi-head classification and loss.

New Features:
- Support multi-class, multi-output classification tasks.
- Introduce `MultiHead` class to manage multiple classification heads.
- Implement `MultiLoss` to apply different loss functions to each head.
- Enable optional label mapping in data files and configuration for Wandb logging.

Tests:
- Update tests to cover multi-head classification and loss functionality.

### CODE REVIEW

**Review by robertavram-md — COMMENTED** [review]
Code looks good; I do think you should add example for inference using the video_inference.py - do you have such examples?

Also does the current code enable the saving of validation .csv files with the predictions?

How about sweeps? Sweeps tend to focus on "best_val_loss" but here you have 3 losses in your example. How do you deal with that?
I apologize if I didn't see these aspects.

**Comment by JacquesDelfrate** [comment]
Answering
> Code looks good; I do think you should add example for inference using the video_inference.py - do you have such examples?
> 
> Also does the current code enable the saving of validation .csv files with the predictions?
> 
> How about sweeps? Sweeps tend to focus on "best_val_loss" but here you have 3 losses in your example. How do you deal with that? I apologize if I didn't see these aspects.


# Adding an Example for Inference Using `video_inference.py`:
I haven't touch that file yet. It is probably broken at the moment. I am going to investigate possible break in `video_inference.py` in a next PR.

# Saving Validation `.csv` Files with Predictions:
Yes, the current implementation supports saving validation .csv files containing the logits predictions (float) and targets (int) for each head and class. 

# Handling Sweeps with Multiple Losses:
The `MultiLoss` class's `forward()` method returns a dictionary containing individual loss values for each head, as well as an aggregated loss (main_loss) which is the sum of all head losses. Here's how it works:

**Individual Losses:** Each head's loss (e.g., contrast_agent, main_structure, stent_presence) is calculated separately. These losses are primarily used for logging purposes with wandb.

**Aggregated Loss (`main_loss`):** During training, the model optimizes using the aggregated loss, which is the sum of all individual losses weighted by their respective head_weights. This aggregated loss is used to determine best_val_loss for sweeps and is also logged with wandb.

**Head Weights Configuration:** The importance of each loss can be controlled via the head_weights in the configuration file (not mentioned above). By default, all heads have a weight of 1.0, but these can be adjusted as needed. Here is an example configuration:

``` yaml
# Head weights configuration
head_weights:
  contrast_agent: !!float 1.0
  main_structure: !!float 1.0
  stent_presence: !!float 1.0
```
**Sweeps Integration:** When conducting sweeps, the best_val_loss is based on the aggregated main_loss. This ensures that the sweep focuses on optimizing the overall performance considering all three losses.
